### DISCUSSION TITLE
Guest portal authentication not working

### DISCUSSION THREAD

**Discussion by LuDomMedia** [Art-of-WiFi/UniFi-API-client#101] [Q&A] [answered]
Hey, I have a problem with the authentication:

I've tried to create an external captive portal with the API, the auth_guest_basic.php file as template (and config.php) and the following settings:

 - Guest Hotspot
 - Guest Portal enabled
 - Authentication Type "External Portal Server"
 - IP of local portal page
 - Local Account with super admin as login account

My Setup:
 - UDM Pro (Firmware Version 1.8.6.2969, UniFi Network Version 6.0.43)
 - Multiple UAC AP Pro's (Firmware Version 4.3.28.11361)
 - Portal Server running on a Raspberry Pi 3 B+

Now my problem: When the guest is redirected to the portal, the output is "false".

I've already tried to google the problem. I saw [a post](https://community.ui.com/questions/Unifi-External-Portal-does-not-authorize/6a56a9f8-6def-4e44-953a-b0b539012011) and tried the code from there. Exactly the same code.
And in Debug-Mode, I see following:

```
-----------LOGIN-------------
Array
(
    [url] => 192.168.0.1/api/login
    [content_type] => text/html; charset=utf-8
    [http_code] => 302
    [header_size] => 448
    [request_size] => 226
    [filetime] => -1
    [ssl_verify_result] => 0
    [redirect_count] => 0
    [total_time] => 0.002616
    [namelookup_time] => 0.000122
    [connect_time] => 0.000123
    [pretransfer_time] => 0.000324
    [size_upload] => 54
    [size_download] => 69
    [speed_download] => 34500
    [speed_upload] => 27000
    [download_content_length] => 69
    [upload_content_length] => 54
    [starttransfer_time] => 0.002487
    [redirect_time] => 0
    [redirect_url] => https://192.168.0.1/
    [primary_ip] => 192.168.0.1
    [certinfo] => Array
        (
        )

    [primary_port] => 80
    [local_ip] => 192.168.0.156
    [local_port] => 51512
    [http_version] => 2
    [protocol] => 1
    [ssl_verifyresult] => 0
    [scheme] => HTTP
    [appconnect_time_us] => 124
    [connect_time_us] => 123
    [namelookup_time_us] => 122
    [pretransfer_time_us] => 324
    [redirect_time_us] => 0
    [starttransfer_time_us] => 2487
    [total_time_us] => 2616
)

----------RESPONSE-----------
HTTP/1.1 302 Found
Vary: Origin
X-DNS-Prefetch-Control: off
X-Frame-Options: SAMEORIGIN
Strict-Transport-Security: max-age=15552000; includeSubDomains
X-Download-Options: noopen
X-Content-Type-Options: nosniff
X-XSS-Protection: 1; mode=block
Accept-Ranges: bytes
Location: https://192.168.0.1
Content-Type: text/html; charset=utf-8
Content-Length: 69
X-Response-Time: 1ms
Date: Sat, 06 Mar 2021 16:15:29 GMT
Connection: keep-alive

Redirecting to https://192.168.0.1.
-----------------------------
false
Connecting to the network...
```

I hope anyone could help me.

Thanks
LuDom Media

**Comment by malle-pietje** [comment]
Please provide some more details such the contents of the config.php file (without the password...) and how you are capturing the device and AP MAC addresses.

**Reply by LuDomMedia** [reply]
Okay, here are the files:

config.php:
```
<?php
/**
 * Copyright (c) 2021, Art of WiFi
 *
 * This file is subject to the MIT license that is bundled with this package in the file LICENSE.md
 */

/**
 * Controller configuration
 * ===============================
 * Copy this file to your working directory, rename it to config.php and update the section below with your UniFi
 * controller details and credentials
 */
$controlleruser     = '************************'; // the user name for access to the UniFi Controller
$controllerpassword = '***********'; // the password for access to the UniFi Controller
$controllerurl      = '192.168.0.1'; // full url to the UniFi Controller, eg. 'https://22.22.11.11:8443', for UniFi OS-based
                          // controllers a port suffix isn't required, no trailing slashes should be added
$controllerversion  = '6.0.43'; // the version of the Controller software, e.g. '4.6.6' (must be at least 4.0.0)

/**
 * set to true (without quotes) to enable debug output to the browser and the PHP error log
 */
$debug = true;
```

index.php:
```
<?php

//PHP SCRIPT FOR SIMPLE PORTAL
//REQUIREMENTS:
//curl needs to be enabled (php_curl.dll for widnows)

session_start();

$_SESSION['id'] = $_GET['id'];          //user's mac address

?>
<!DOCTYPE html>
<html>
<head>
    <meta charset="utf-8">
    <title>Portal Page Example</title>
</head>

<body>
    <form name="login" action="authorized.php" method="post">
    <input id="submit" type="submit" name="submit" value="Connect" />
    </form>
</body>
</html>
```

authorized.php:
```
<?php

//PHP SCRIPT FOR SIMPLE PORTAL
//REQUIREMENTS:
//curl needs to be enabled (php_curl.dll for widnows)

session_start();

$_SESSION['id'] = $_GET['id'];          //user's mac address

?>
<!DOCTYPE html>
<html>
<head>
    <meta charset="utf-8">
    <title>Portal Page Example</title>
</head>

<body>
    <form name="login" action="authorized.php" method="post">
    <input id="submit" type="submit" name="submit" value="Connect" />
    </form>
</body>
</html>
```

**Comment by malle-pietje** [comment, marked as answer]
Ok, to start with the controller URL is missing the **https://** prefix.

**Comment by LuDomMedia** [comment]
Thank you! It worked

**Reply by malle-pietje** [reply]
You’re welcome!

**Comment by simonriv** [comment]
Hi, I have a UDM PRO in which I am configuring a Hostpot Portal, in the redirection settings of the UDM I configured an external server, this external server is an API that I configure with Node JS to receive the connection requests to then show the user a form where he registers his data, up to this point everything works correctly, the problem I am having is when authorizing the user on the network to browse the internet I don't know if I should send any http response to the UDM then I need help with this.

The connection request that arrives to the external server when someone connects looks like this:

![image2](https://github.com/Art-of-WiFi/UniFi-API-client/assets/85804417/8f98dfdc-fbfe-4b9e-9045-e5c61859ddc3)


How do I authorize the user or what response should I send to allow the connection?

**Reply by malle-pietje** [reply]
This question is not really related to the API client. It is more about how to build a captive portal.

I will leave this comment here, maybe someone can contribute some suggestions.
